velox/functions/prestosql/Repeat.cpp

```diff
@@ -0,0 +1,136 @@
+ * limitations under the License.
+ */
+#include "velox/expression/DecodedArgs.h"
+#include "velox/expression/VectorFunction.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+// See documentation at https://prestodb.io/docs/current/functions/array.html
+class RepeatFunction : public exec::VectorFunction {
+ public:
+  static constexpr int32_t kMaxResultEntries = 10000;
+
+  bool isDefaultNullBehavior() const override {
+    // For repeat(null, count), it should output an array with
+    // repeating null for "count" times.
+    return false;
+  }
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    exec::DecodedArgs decodedArgs(rows, args, context);
+
+    auto elementVector = args[0];
+    auto countDecoded = decodedArgs.at(1);
+
+    // Calculate the total number of the new elements.
+    auto totalCount = 0;
+    rows.applyToSelected([&](vector_size_t row) {
+      auto count =
+          countDecoded->isNullAt(row) ? 0 : countDecoded->valueAt<int32_t>(row);
+      VELOX_USER_CHECK_GE(
+          count,
+          0,
+          "Count argument of repeat function must be greater than or equal to 0");
+      VELOX_USER_CHECK_LE(
+          count,
+          kMaxResultEntries,
+          "Count argument of repeat function must be less than or equal to 10000")
+      totalCount += count;
+    });
+
+    auto numRows = rows.size();
+    memory::MemoryPool* pool = context.pool();
+
+    // Allocate new vectors for nulls if necessary.
+    BufferPtr dictNulls;
+    uint64_t* rawDictNulls = nullptr;
+    if (elementVector->mayHaveNulls()) {
+      dictNulls = allocateNulls(totalCount, pool);
+      rawDictNulls = dictNulls->asMutable<uint64_t>();
+    }
+    BufferPtr arrayNulls;
+    uint64_t* rawArrayNulls = nullptr;
+    if (countDecoded->mayHaveNulls()) {
+      arrayNulls = allocateNulls(numRows, pool);
+      rawArrayNulls = arrayNulls->asMutable<uint64_t>();
+    }
+
+    // Allocate new vectors for indices, length and offsets.
+    BufferPtr indices = allocateIndices(totalCount, pool);
+    BufferPtr sizes = allocateSizes(numRows, pool);
+    BufferPtr offsets = allocateOffsets(numRows, pool);
+    auto rawIndices = indices->asMutable<vector_size_t>();
+    auto rawSizes = sizes->asMutable<vector_size_t>();
+    auto rawOffsets = offsets->asMutable<vector_size_t>();
+
+    vector_size_t offset = 0;
+    rows.applyToSelected([&](vector_size_t row) {
+      if (rawArrayNulls && countDecoded->isNullAt(row)) {
+        // If count is null, the result should be null.
+        // e.g. repeat('a', null) -> null
+        bits::setNull(rawArrayNulls, row);
+        return;
+      }
+
+      auto count = countDecoded->valueAt<int32_t>(row);
+      rawSizes[row] = count;
+      rawOffsets[row] = offset;
+      for (auto i = 0; i < count; i++) {
+        rawIndices[offset] = row;
```

Comment:
use std::fill to populate offsets

Reply:
Could you please elaborate this a bit? How do we use std::fill given that `offsets` is not a vector? Searched in the codebase but not able to find similar usage.

Reply:
std::fill doesn't require a vector: https://en.cppreference.com/w/cpp/algorithm/fill

Here is an example:

```
  BufferPtr indices = allocateIndices(10, pool());
  auto rawIndices = indices->asMutable<vector_size_t>();
  std::fill(rawIndices + 3, rawIndices + 7, 33);
  LOG(ERROR) << printIndices(indices);

2 unique indices out of 10: 0, 0, 0, 33, 33, 33, 33, 0, 0, 0
```